### PR TITLE
bootloader: only print "failed to execute" error on non-SystemExit errors

### LINKED ISSUE #4860: New - and consistent - failiures of the QtQml tests on macOS (CI)
## Context information

* Output of `pyinstaller --version`: develop.
* Version of Python: e.g. 3.7.5 
* Platform: OS X (macOS CI on Travis)

### Issue
QtQml is consistently failing. From Travis:
 
```
----------------------------- Captured stdout call -----------------------------
------- Starting build. -------
------- Build finshed, now running executable. -------
[7439]  RUNNING:  '/private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_Qt5_QtQml_onedir_PyQt5_0/dist/test_source/test_source' , args:  ['./test_source']
----------------------------- Captured stderr call -----------------------------
------- Starting build. -------
------- Build finshed, now running executable. -------
[7439] PyInstaller Bootloader 3.x
[7439] LOADER: executable is /private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_Qt5_QtQml_onedir_PyQt5_0/dist/test_source/test_source
[7439] LOADER: homepath is /private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_Qt5_QtQml_onedir_PyQt5_0/dist/test_source
[7439] LOADER: _MEIPASS2 is NULL
[7439] LOADER: archivename is /private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_Qt5_QtQml_onedir_PyQt5_0/dist/test_source/test_source
[7439] LOADER: No need to extract files to run; setting extractionpath to homepath
[7439] LOADER: Already in the child - running user's code.
[7439] LOADER: Python library: /private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_Qt5_QtQml_onedir_PyQt5_0/dist/test_source/Python
[7439] LOADER: Loaded functions from Python library.
[7439] LOADER: Manipulating environment (sys.path, sys.prefix)
[7439] LOADER: sys.prefix is /private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_Qt5_QtQml_onedir_PyQt5_0/dist/test_source
[7439] LOADER: Pre-init sys.path is /private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_Qt5_QtQml_onedir_PyQt5_0/dist/test_source/base_library.zip:/private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_Qt5_QtQml_onedir_PyQt5_0/dist/test_source
[7439] LOADER: Setting runtime options
[7439] LOADER: Initializing python
[7439] LOADER: Overriding Python's sys.path
[7439] LOADER: Post-init sys.path is /private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_Qt5_QtQml_onedir_PyQt5_0/dist/test_source/base_library.zip:/private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_Qt5_QtQml_onedir_PyQt5_0/dist/test_source
[7439] LOADER: Setting sys.argv
[7439] LOADER: setting sys._MEIPASS
[7439] LOADER: importing modules from CArchive
[7439] LOADER: extracted struct
[7439] LOADER: callfunction returned...
[7439] LOADER: extracted pyimod01_os_path
[7439] LOADER: callfunction returned...
[7439] LOADER: extracted pyimod02_archive
[7439] LOADER: callfunction returned...
[7439] LOADER: extracted pyimod03_importers
[7439] LOADER: callfunction returned...
[7439] LOADER: Installing PYZ archive with Python modules.
[7439] LOADER: PYZ archive: PYZ-09.pyz
[7439] LOADER: Running pyiboot01_bootstrap.py
[7439] LOADER: Running pyi_rth_pyqt5.py
[7439] LOADER: Running test_source.py
[7439] Failed to execute script test_source
[7439] Error: 0
[7439] Traceback: ['Traceback (most recent call last):\n', '  File "/private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_Qt5_QtQml_onedir_PyQt5_0/test_source.py", line 33, in <module>\n    sys.exit(res)\n', 'int: 0\n']
[7439] LOADER: OK.
[7439] LOADER: Cleaning up Python interpreter.
------------------------------ Captured log call -------------------------------
```

### Links to failing builds:

First noticed in https://travis-ci.org/github/bjones1/pyinstaller/jobs/683086293

* https://travis-ci.org/github/pyinstaller/pyinstaller/jobs/683275309
* https://travis-ci.org/github/pyinstaller/pyinstaller/jobs/683703566
* https://travis-ci.org/github/bjones1/pyinstaller/jobs/683551334

## Solution

I'd suggest marking these tests as `skipif(is_darwin, reason='Issue #4860')` temporarily whilst we find a solution to allow CI to keep passing.

### PR DESCRIPTION
**Pull request by Legorooj**
#4592 added a new traceback to the bootloader. That caused an error that was originally reported and fixed in #1869 and 36b6ab3. The error was that `SystemExit` - the standard script exit exception, raised by `sys.exit` and other exit functions, was handled as a normal exception, and the bootloader printed `Failed to execute script X`, before erroring.

This PR fixes this by checking if this error is a SystemExit _first_.

Fixes #4860

### CODE REVIEW

**Comment by Legorooj** [comment]
Stuff to do:

- [ ] Move the FATALERROR call to after `PI_PyErr_Print` call when not using windowed mode
- [ ] Check if the exception is a SystemExit when in windowed mode, and if so, handle it gracefully before exiting.

File: bootloader/src/pyi_python.c [hunk 1]
```diff
@@ -122,12 +134,15 @@ pyi_python_map_names(HMODULE dll, int pyvers)
...
+    GETPROC(dll, PyErr_GivenExceptionMatches);
+    GETPROC(dll, Py_Exit);
     GETPROC(dll, PyImport_AddModule);
     GETPROC(dll, PyImport_ExecCodeModule);
     GETPROC(dll, PyImport_ImportModule);
     GETPROC(dll, PyList_Append);
     GETPROC(dll, PyList_New);
     GETPROC(dll, PyLong_AsLong);
+    // GETPROC(dll, PyLong_Check);
     GETPROC(dll, PyModule_GetDict);
```
Comment (Legorooj) [inline bootloader/src/pyi_python.c:146]: If I add this in, the bootloader doesn't run. Any ideas? Same for the other commented out line above.

File: bootloader/src/pyi_launch.c [hunk 2]
```diff
@@ -358,6 +358,46 @@ pyi_launch_extract_binaries(ARCHIVE_STATUS *archive_status)
...
 
+// #if defined(LAUNCH_DEBUG) && defined(WINDOWED)
+static void 
+_handle_system_exit(PyObject *exception, PyObject *value, PyObject *tb) {
+    int exitcode = 0;
+    
+    FATALERROR("fetching exception");
+    PI_PyErr_Fetch(&exception, &value, &tb);
+    FATALERROR("exception fetched");
+    if (!PI_PyErr_GivenExceptionMatches(exception, PI_PyExc_SystemExit)) {
```
Comment (Legorooj) [inline bootloader/src/pyi_launch.c:369]: @danyeaw @bjones1 @htgoebel you're better at C than I am. I've stuck FATALERROR calls everywhere to debug, and the bootloader just crashes on this line.

I'm building with the following command: `pyinstaller test.py --onefile -d bootloader --clean -w`

And test.py is:
```python
raise SystemExit(13)
```

File: bootloader/src/pyi_launch.c [hunk 3]
```diff
@@ -358,6 +358,46 @@ pyi_launch_extract_binaries(ARCHIVE_STATUS *archive_status)
...
+    if (value == NULL || value == PI_Py_None) {
+        FATALERROR("Exiting with code: %i\n", exitcode);
+        goto done;
+    }
+    if (PI_PyLong_Check(value)) {
+        exitcode = (int) PI_PyLong_AsLong(value);
+        FATALERROR("Exiting with code: %i\n", exitcode);
+    } else {
+        PyObject* code = PI_PyObject_GetAttrString(value, "code");
+        exitcode = (int) PI_PyLong_AsLong(code);
```
Comment (htgoebel) [inline bootloader/src/pyi_launch.c:384]: What is this case? Please add a short comment.
This will fail if value doe not have an attribute "code".

**Comment by htgoebel** [comment]
Another idea to simplify the code: The reason for the added `#if windows …` code ist that in --windowed mode sys.stderr is NullWriter, which will will inhibit printing the traceback. (see https://github.com/pyinstaller/pyinstaller/pull/4592#issuecomment-616114486). So why not simply do something like this:
``` C
// scratch
stderr = PySys_GetObject("__stderr__");
PySys_SetObject("stderr", stderr);
PI_PyErr_Print();
```
This of course requires `sys.__stderr__` not being NULL in windowed mode.

**Comment by htgoebel** [comment]
:-( This will not work, as `__stderr__` will be None, too, see https://docs.python.org/3/library/sys.html?highlight=__stderr__#sys.__stderr__
> Under some conditions stdin, stdout and stderr as well as the original values `__stdin__`, `__stdout__` and `__stderr__` can be None. It is usually the case for Windows GUI apps that aren’t connected to a console and Python apps started with pythonw.

**Comment by Legorooj** [comment]
@htgoebel I had an idea last night :-). I've moved the `PI_PyErr_Print` call to before the windowed-only traceback like:
```c
PI_PyErr_Print();
FATALERROR("Failed to execute...");

#if defined(WINDOWED) && defined(LAUNCH_DEBUG)
// Windowed traceback
#endif
```

My rationale for this change is that when running in windowed mode the `PI_PyErr_Print` call won't do anything _except_ handle a SystemExit gracefully, because `stderr` etc are `NULL`. If we aren't running in windowed mode, then this will work just fine. What do you think? I think this is a lot more graceful than the other method I was using.

**Comment by htgoebel** [comment]
Top!

**Pull request merged**